### PR TITLE
Add optional `ignore_edge_order` to contractors. #315

### PR DESCRIPTION
**Pull request by benjaminr**
Addresses #315

### CODE REVIEW

File: tensornetwork/contractors/opt_einsum_paths/path_contractors.py [hunk 1]
```diff
@@ -171,18 +189,20 @@ def optimal(
...
     The network after full contraction.
   """
   alg = functools.partial(opt_einsum.paths.optimal, memory_limit=memory_limit)
-  return base(nodes, alg, output_edge_order)
+  return base(nodes, alg, output_edge_order, ignore_edge_order)
 
 
 def branch(nodes: Union[TensorNetwork, Iterable[BaseNode]],
-           output_edge_order: Optional[Sequence[Edge]] = None,
+           output_edge_order: Sequence[Edge] = None,
```
Comment (Thenerdstation) [inline tensornetwork/contractors/opt_einsum_paths/path_contractors.py]: Don't remove the `Optional` keyword here.

File: tensornetwork/contractors/opt_einsum_paths/path_contractors.py [hunk 2]
```diff
@@ -202,19 +222,21 @@ def branch(nodes: Union[TensorNetwork, Iterable[BaseNode]],
...
       opt_einsum.paths.branch, memory_limit=memory_limit, nbranch=nbranch)
-  return base(nodes, alg, output_edge_order)
+  return base(nodes, alg, output_edge_order, ignore_edge_order)
 
 
 def greedy(
     nodes: Union[TensorNetwork, Iterable[BaseNode]],
-    output_edge_order: Optional[Sequence[Edge]] = None,
-    memory_limit: Optional[int] = None):  # -> Union[BaseNode, TensorNetwork]:
+    output_edge_order: Sequence[Edge] = None,
```
Comment (Thenerdstation) [inline tensornetwork/contractors/opt_einsum_paths/path_contractors.py]: Don't remove `Optional` here.

File: tensornetwork/contractors/opt_einsum_paths/path_contractors.py (hunk 1 above)
Comment (benjaminr) [reply tensornetwork/contractors/opt_einsum_paths/path_contractors.py]: Sorry think this is just where my merge went wonky.

File: tensornetwork/contractors/opt_einsum_paths/path_contractors.py (hunk 2 above)
Comment (benjaminr) [reply tensornetwork/contractors/opt_einsum_paths/path_contractors.py]: Sorry think this is just where my merge went wonky.

File: tensornetwork/contractors/opt_einsum_paths/path_contractors_test.py [hunk 3]
```diff
@@ -92,3 +92,21 @@ def __call__(self, inputs, output, size_dict, memory_limit=None):
...
+  a[0] ^ b[0]
+  a[1] ^ b[1]
+  a[2] ^ b[2]
+
+  e0 = b[3]
+  e1 = b[4]
+
+  final_node = path_algorithm(net,
+                              [e1, e0],
+                              ignore_edge_order=True).get_final_node()
```
Comment (Thenerdstation) [inline tensornetwork/contractors/opt_einsum_paths/path_contractors_test.py]: Seems kinda pointless to test `ignore_edge_order=True` when an edge order is given anyway. :)

Comment (Thenerdstation) [inline tensornetwork/contractors/opt_einsum_paths/path_contractors_test.py]: Please move this test to `path_contractors_node_test` and don't use the `TensorNetwork` object as it is getting nuked in #333

Comment (benjaminr) [reply tensornetwork/contractors/opt_einsum_paths/path_contractors_test.py]: I see what you mean, but it also ensures it's effective despite being passed an order, which I think is a valid test case.

Comment (Thenerdstation) [reply tensornetwork/contractors/opt_einsum_paths/path_contractors_test.py]: Can you add another similar test where you don't include the list of edges?

Comment (benjaminr) [reply tensornetwork/contractors/opt_einsum_paths/path_contractors_test.py]: Sure thing, I'll add to path_contractors_node_test as detailed below.

File: tensornetwork/contractors/opt_einsum_paths/path_contractors_node_test.py [hunk 4]
```diff
@@ -142,3 +142,33 @@ def test_single_node_reorder(backend, path_algorithm):
...
+  a = tn.Node(np.ones((1, 1, 1)), backend=backend)
+  b = tn.Node(np.ones((1, 1, 1, 2, 3)), backend=backend)
+
+  a[0] ^ b[0]
+  a[1] ^ b[1]
+  a[2] ^ b[2]
+
+  final_node = path_algorithm({a, b},
+                              ignore_edge_order=True)
+  assert final_node.tensor.shape == (2, 3)
```
Comment (Thenerdstation) [inline tensornetwork/contractors/opt_einsum_paths/path_contractors_node_test.py]: It's unlikely that the final shape will remain consistent with `ignore_edge_order=True`. Can you instead test that `set(final_nodes.edges) == {e[0], e[1]}` (same for below.)

**Comment by Thenerdstation** [comment]
Great work! Thank you for your contribution!

**Comment by benjaminr** [comment]
> Great work! Thank you for your contribution!

Thanks @Thenerdstation!

**Pull request merged**